### ISSUE TITLE
Drawing a lot of lines of different color

### ISSUE THREAD

**Issue by ipfreely42** [helix-toolkit/helix-toolkit#439]
Hi everyone, apologies if this is not the right place to post my question.

I am using HelixToolkit.SharpDX. Based on a list of points, I need to draw a grid consisting of a potentially large number of lines. So far I have been using the LineGeometryModel3D which yields very good results. However, instead of defining a global color value, I need to be able to set a different color for each line segment.

Is there any way of achieving this while maintaining a reasonable performance?

I am rather new to 3D programming and am more or less pushed into this field, so any help would be appreciate heaps.

**Comment by holance** [comment]
Simple way is to use several LineGeometryModel3D with different color to draw.

**Comment by ipfreely42** [comment]
Thanks, but I already tried this. Unfortunately, the performance is not good enough.

**Comment by holance** [comment]
How many color do you need? Try to merge all the lines with same color in same line geometry model to reduce the draw calls. Or try to only draw the lines when visible in the camera.

**Comment by holance** [comment]
If you don't care the line thickness. Set cullmode = wireframe, thickness = 0 to increase performance.

**Comment by ipfreely42** [comment]
I had a look at the HelixToolkit.Wpf examples and found the RangeColorAxis class. This is pretty much what I need. Is there a similar way of achieving this with HelixToolkit.SharpDX? Otherwise I'd try to port the class to the project and see where this would get me. However, as stated above, I am pretty new to 3D programming.

**Comment by holance** [comment]
Seems like you might need to write your own shader to achieve what you want.

**Comment by ipfreely42** [comment]
Hm, ok, thanks for that input. I have no idea where to start, though, as I have never done shader programming before. I guess I'll start by going through the examples.

Thanks.

**Comment by smischke** [comment]
If you modify "Examples\WPF.SharpDX\LineShadingDemo\MainViewModel.cs" slightly...

```
            // lines model3d
            var e1 = new LineBuilder();
            e1.AddBox(new Vector3(0, 0, 0), 1, 0.5, 2);
            //e1.AddLine(new Vector3(-1, 0, 0), new Vector3(1, 0, 0));
            this.Lines = e1.ToLineGeometry3D();
// >>
            this.Lines.Colors = new Color4Collection();
            foreach (var line in this.Lines.Lines)
            {
                this.Lines.Colors.Add(new Color4(line.P0));
            }
// <<
```

... you should get the desired result.

**Comment by holance** [comment]
Fixed in #452
